File: ts/src/common/http.ts
import { createHash } from "crypto";
import { createWriteStream, mkdirSync } from "fs";
import { dirname } from "path";
import { Readable } from "stream";
import { pipeline } from "stream/promises";

const USER_AGENT =
  "security-research-bot/1.0 (plugin artifact scanner; contact security@example.com)";

const RETRY_STATUS_CODES = new Set([429, 500, 502, 503, 504]);

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class RateLimitedClient {
  private lastRequest = 0;
  private readonly minInterval: number;

  constructor(requestsPerSecond = 1.0) {
    this.minInterval = 1000 / requestsPerSecond;
  }

  private async wait(): Promise<void> {
    const elapsed = Date.now() - this.lastRequest;
    if (elapsed < this.minInterval) {
      await sleep(this.minInterval - elapsed);
    }
    this.lastRequest = Date.now();
  }

  async get(url: string, options: RequestInit = {}): Promise<Response> {
    return this.withRetry(async () => {
      await this.wait();
      const response = await fetch(url, {
        ...options,
        signal: AbortSignal.timeout(30_000),
        headers: {
          "User-Agent": USER_AGENT,
          ...(options.headers as Record<string, string> | undefined),
        },
      });
      if (!response.ok) {
        throw Object.assign(
          new Error(`HTTP ${response.status} for ${url}`),
          { status: response.status }
        );
      }
      return response;
    });
  }

  async getJson<T>(url: string, options: RequestInit = {}): Promise<T> {
    const response = await this.get(url, options);
    return response.json() as Promise<T>;
  }

  /**
   * Stream-download url to destPath. Returns sha256 hex digest.
   */
  async download(url: string, destPath: string): Promise<string> {
    return this.withRetry(async () => {
      await this.wait();
      const response = await fetch(url, {
        signal: AbortSignal.timeout(120_000),
        headers: { "User-Agent": USER_AGENT },
      });
      if (!response.ok) {
        throw Object.assign(
          new Error(`HTTP ${response.status} downloading ${url}`),
          { status: response.status }
        );
      }
      if (!response.body) {
        throw new Error(`No response body for ${url}`);
      }

      mkdirSync(dirname(destPath), { recursive: true });

      const hash = createHash("sha256");
      const dest = createWriteStream(destPath);

      // Convert web ReadableStream -> Node Readable, tee through hash
      const nodeStream = Readable.fromWeb(
        response.body as import("stream/web").ReadableStream
      );

      await pipeline(nodeStream, async function* (source) {
        for await (const chunk of source) {
          hash.update(chunk as Buffer);
          yield chunk;
        }
      }, dest);

      return hash.digest("hex");
    });
  }

  private async withRetry<T>(
    fn: () => Promise<T>,
    maxRetries = 3
  ): Promise<T> {
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        return await fn();
      } catch (err: unknown) {
        const status =
          err instanceof Error && "status" in err
            ? (err as { status: number }).status
            : undefined;
        const isRetryable =
          status !== undefined && RETRY_STATUS_CODES.has(status);
        if (attempt === maxRetries || !isRetryable) throw err;
        const backoff = Math.pow(2, attempt) * 1000;
        await sleep(backoff);
      }
    }
    throw new Error("unreachable");
  }
}

File: common/http.py
import hashlib
import time
from pathlib import Path

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


class RateLimitedSession:
    """HTTP session with per-request rate limiting and automatic retries."""

    def __init__(self, requests_per_second: float = 1.0):
        self._min_interval = 1.0 / requests_per_second
        self._last_request = 0.0
        self._session = requests.Session()
        retry = Retry(
            total=3,
            backoff_factor=2,
            status_forcelist=[429, 500, 502, 503, 504],
        )
        adapter = HTTPAdapter(max_retries=retry)
        self._session.mount("https://", adapter)
        self._session.mount("http://", adapter)
        self._session.headers["User-Agent"] = (
            "security-research-bot/1.0 (plugin artifact scanner)"
        )

    def _wait(self):
        elapsed = time.time() - self._last_request
        if elapsed < self._min_interval:
            time.sleep(self._min_interval - elapsed)
        self._last_request = time.time()

    def get(self, url: str, **kwargs) -> requests.Response:
        self._wait()
        resp = self._session.get(url, timeout=30, **kwargs)
        resp.raise_for_status()
        return resp

    def download(self, url: str, dest_path: str) -> str:
        """Stream-download url to dest_path. Returns sha256 hex digest."""
        self._wait()
        resp = self._session.get(url, timeout=120, stream=True)
        resp.raise_for_status()
        Path(dest_path).parent.mkdir(parents=True, exist_ok=True)
        hasher = hashlib.sha256()
        with open(dest_path, "wb") as f:
            for chunk in resp.iter_content(chunk_size=65536):
                f.write(chunk)
                hasher.update(chunk)
        return hasher.hexdigest()
